### ISSUE TITLE
Any sense in separating mapping annotations from the rest of the API?

### ISSUE THREAD

**Issue by rbygrave** [eclipse-ee4j/jpa-api#246]
So given the package change to `jakarta.persistence` I wonder if this is an opportunity to separate the mapping annotations:
- from the rest of the API as a separate artifact?
- in the same artifact but under a mapping package?

That is part of me wonders if in moving forward it would be useful to separate the mapping annotations such that mapping changes / improvements could be released more easily / frequently?  (It might be easier to get agreement on mapping improvements separately from other API improvements?).

Just a thought but given the package change it seems like an opportunity to consider it.

(Maybe this has been discussed in email so I'll go and have a look / search - apologies if it's already been discussed).

Cheers, Rob.

**Comment by m-reza-rahman** [comment]
I don't think this is possible now without breaking backwards compatibility. I suggest just closing this to reduce clutter and be able to focus on higher priority issues.

Reza Rahman
Jakarta EE Ambassador, Author, Blogger, Speaker

Please note views expressed here are my own as an individual community member and do not reflect the views of my employer.

**Comment by m-reza-rahman** [comment]
I don't think this is possible now without breaking backwards compatibility. I suggest just closing this to reduce clutter and be able to focus on higher priority issues.

Reza Rahman
Jakarta EE Ambassador, Author, Blogger, Speaker

Please note views expressed here are my own as an individual community member and do not reflect the views of my employer.

**Comment by gavinking** [comment]
> * from the rest of the API as a separate artifact?
> * in the same artifact but under a mapping package?

I think it would be _very much_ worth doing both these things.

It was a huge mistake to have put the mapping annotations in the same package as the APIs, and it's well past time to own up to that mistake and correct it.

**Comment by lukasj** [comment]
> It was a huge mistake to have put the mapping annotations in the same package as the APIs

agree, I've set this up for consideration for the next major version

**Comment by gavinking** [comment]
> agree, I've set this up for consideration for the next major version

Excellent!

**Comment by victorwss** [comment]
Is the idea to separate the mapping annotations on one side (along with some enums) and the `EntityManager`, `Query` and friends, criteria, metamodel and SPI in the other side (possibly also broken down in smaller pieces)? If so, I think that it isn't a good idea and would only create more backward-compatibility and JAR-hell problems for little to no benefits whatsoever.

The migration from javax to jakarta already does great damage and headache to JPA's users, severely breaking the "write once, run everywhere" that allowed people to trust Java. It punishes users and Java developers everywhere preventing, delaying, complicating and/or screwing up software evolution just for the sake of a legalese nitpick problem about naming rights that in fact was created many years ago by Sun and that users and developers should not have to ever care about, but unfortunately have to do. So please, we should not ever again break backward-compatibility without an extremely good reason and I see nothing like that here.

So, what are the real benefits that end users would gain from this separation?

**Comment by rbygrave** [comment]
> what are the real benefits that end users would gain from this separation?

As I see it, the idea is that this would enable other JPA API's to be developed where the mapping is fully reused but the API could be different. For example, perhaps more like Panache or perhaps more along the lines of "Sessionless EntityManager" or maybe something else even.

### Panache
With for example something like Panache - https://quarkus.io/guides/hibernate-orm-panache  we see the entity mapping annotations reused etc but we don't see EntityManager as exposed API.

### "Stateless EntityManager"
I'm extrapolating a **_LOT_** on Gavins suggestion here ... but taking the idea of "Stateless EntityManager" in https://github.com/jakartaee/persistence/issues/374  it could be that there could be a "Stateless JPA API" that could be drafted that reuses all the entity mapping but provides a "Stateless EntityManager API" which might be functionally along the lines of Hibernate StatelessSession. Would it be easier for that to develop as a separate API?

Maybe the separation of the mapping annotations would enable a "lightweight" version of JPA to develop which could be argued is more suited for use in "smaller" apps like Lambdas.

Disclosure: I maintain an ORM that uses the JPA mapping annotations but does not use the API (EntityManager etc) so my bias is that I'd prefer the mapping to be separate.

**Comment by gavinking** [comment]
We need to make a decision on this one.

It would be wonderful to have the mapping annotations in a separate package. 

On the other hand, the correct moment to do this would have been when we moved to the `jakarta.persistence` namespace, and it was a huge missed opportunity to not have done it at that time.

Doing it now is, I would say, probably not worth the disruption it's going to cause.

**Comment by sebersole** [comment]
-1 from me.  Just does not seem worth the disruption.

**Comment by AleksNo** [comment]
With an updated Eclipse Transformer the disruption could be minimal.

**Comment by gavinking** [comment]
> Just does not seem worth the disruption.

I agree, but let's just suspend disbelief for a bit and see what it would look like. How might we go about this?

Option 1: rip off the bandaid
-----------------------------

Simply move all the annotations into a new package (actually two new packages because we want to segregate "mapping" annotations from "modeling" annotations.)

This option forces immediate migration.

Option 2: deprecate/replace
----------------------------

Deprecate all the old annotations, and create new ones in the new packages.

This means carrying all the deprecated annotations around for a whole major version. We would have to rush out a Persistence 5.0 right after.

Obviously, having two sets of annotations is much, much worse than just having just one set, so Persistence 4.0 would be *worse* than Persistence 3.2.

Option 3: split the Persistence module
--------------------------------------

Move the annotations into a new `jakarta-persistence-annotations.jar`. Now or later (4.1?), introduce a second version of the `jar`, `jakarta-persistence-annotations2.jar`, or whatever, and require providers to support both jars until at least the next major version.

This causes little immediate disruption to application developers.

On the other hand, it's a significant burden on implementors. It's also at least a bit painful for platforms like Quarkus. Not at _runtime_, since the two jars can safely coexist at runtime, but at development time; but it's probably manageable.

A wrinkle here is that you would probably need both `jars` as *runtime* dependencies, even if only one of them is a compile-time dependency (of the application). But I suppose that's OK?

Migration
---------

In all of these the migration story is similar: the developer has to do a quite messy text search/replace on each individual annotation name, since not every `import jakarta.persistence` statement is changing, and not every annotation is moving to the same package. Perhaps by the time this goes out this is the sort of thing that coding assistants will do reliably, but Junie doesn't seem capable of this kind of thing today. 

On the other hand, it would be easy to write a shell script which performs the migration, and we could do that and make it available to the community. We could even just give them the long-ass regex to copy/paste into their IDE.

Clearly, options 2 and 3 are preferable, since they don't require *immediate* migration.

Bottom line
-----------

The only one that looks possibly acceptable to me is Option 3.

At least, it looks like enough of a possibility that I would be prepared to do the `jar` split now, even if we delay introduction of the second `jar` with the repackaged annotations until 4.1 or possibly never. @rbygrave has already presented completely independent reasoning which could motivate such a split (reuse of the mapping annotations in ORMs which don't implement the Persistence APIs).

@sebersole as an implementor, how hard would it be to do some sort of annotation type->annotation type transformation in `hibernate-models` or wherever? What I mean is, read a `@jakarta.persistence.mapping.Column` and dress it up as a `@jakarta.persistence.Column`.

Thoughts?

**Comment by beikov** [comment]
Supporting annotations from a different/old namespace wouldn't be a big deal for Hibernate ORM, but it would surely cause a lot of problems/discomfort for users of the API. I don't think that this breakage is worth the trouble it would cause.

**Comment by sebersole** [comment]
> [@sebersole](https://github.com/sebersole) as an implementor, how hard would it be to do some sort of annotation type->annotation type transformation in `hibernate-models` or wherever? 

Sure, we *could*.  But again, I just don't see the benefit at this time.

> What I mean is, read a `@jakarta.persistence.mapping.Column` and dress it up as a `@jakarta.persistence.Column`.

Implementation detail, but I'd probably opt to go the other way under the assumption that `@jakarta.persistence.Column` would go away eventually (I mean, if not, I doubly don't see the benefit) - I'd rather just bite the bullet at that time rather than later when we might be pressed on other things.

**Comment by gavinking** [comment]
> it would surely cause a lot of problems/discomfort for users

Do you guys think *just splitting the existing API across two different `jar` files* (as a split package) would cause significant discomfort to users?

I'm not seeing why it would.

And if we did that, then we could completely kick the can down the road with respect to anything else. (Which what I really want to do here.) On the other hand If we *don't* do that `jar` split now, then we may as well just close the issue, because it's definitely never going to happen.

**Comment by beikov** [comment]
I don't think splitting into two JARs is a big deal, but since two Java modules/JARs can't share the same packages, doing that would make running in a modular environment impossible.

**Comment by gavinking** [comment]
> since two Java modules/JARs can't share the same packages, doing that would make running in a modular environment impossible.

Ah, sure, right.

**Comment by gavinking** [comment]
Looks like we're not going to do this. We missed the chance we had.
